Git creation and tracking (#59)
arcus/education_modules: git_creation_and_tracking/git_creation_and_tracking.md

[Moon](https://commons.wikimedia.org/wiki/File:Lune_ico.png)
$\copyright$; Luc Viatour / [https://lucnix.be](https://lucnix.be/) /
[CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/deed.en).

In the examples, we will see code as if it is written by Vlad Dracula. If you are practicing these steps as you move through this module, your name and email address will appear instead of Dracula's.

To copy the code in the examples, enter everything after the dollar sign "$" into the command line, and then press the enter or return key. If the command has any output, it will not start with a dollar sign, but a new line where you can type after a dollar sign will appear.

```console
$logname
vlad
$
```

The **codeblock** above tells you that when Dracula types `logname` and presses enter, the output is 'vlad', his username. Try entering `logname` in your own console. There will be other code in front of each `$` describing who and where you are. Since the stuff before the `$` will be different for everyone, it is customary to omit it from example code. We will also omit the `$` after the output.

### Initializing the repository

> First, let's create a directory in `Desktop` folder for our work and then move into that directory:

Reviewer: Consider adding some instructions for folks who don't have Desktop (e.g. will this work for Windows folks?)

Author: Good call, I changed it to a suggestion and included a suggestion of `C` drive for Windows users.

Next we tell Git to make `planets` a [repository](https://swcarpentry.github.io/git-novice/reference.html#repository) (also called a **repo**)
-- a place where Git can store versions of our files:

```console
$ git init
```

It is important to note that `git init` will create a repository that
includes subdirectories and their files---there is no need to create
separate repositories nested within the `planets` repository, whether
subdirectories are present from the beginning or added later. Also, note
that the creation of the `planets` directory and its initialization as a
repository are completely separate processes.

If we use `ls`, the **list** command, to show the directory's contents,
it appears that nothing has changed:

```console
$ ls
```

Like many commands, `ls` can be modified using **flags**, very short bits of code that change some of the options on the command. If we add the `-a` flag to `ls`, we will be asking Git to list all files.
Now we can see that Git has created a hidden directory within `planets` called `.git`:

```console
$ ls -a

> .	..	.git

Reviewer: what do the dots . and .. mean here?  Consider explaining

Author: good call, added a "learnmore" box!

```

Git uses this special subdirectory to store all the information about the project,
including all files and sub-directories located within the project's directory.
If we ever delete the `.git` subdirectory,
we will lose the project's history.

### Creating the `main` branch

The `main` branch is likely to be the default branch depending on your settings and version
of Git. But just to be sure, we will change the default branch to be called `main`:

> 

Reviewer: add a hint box that explains the existence of "master" which folks *will* see at some point... deets to talk about or link to in https://www.zdnet.com/article/github-to-replace-master-with-alternative-term-to-avoid-slavery-references/

Author: @pm0kjp  Added a box with brief explanation and link, but would love some eyes on it.

>You will see that the primary branch of some projects, particularly older projects, is called `master` instead of `main`. GitHub has joined the programming community in a concerted effort to [replace programming terms associated with human slavery](https://www.zdnet.com/article/github-to-replace-master-with-alternative-term-to-avoid-slavery-references/).

Author: @pm0kjp 

I copied that from earlier modules, so when we decide on a position we should update the two setup modules to reflect it too. More thoughts to come!

Also @rosemm for when we make a conscious decision on this!

fatal:
Not a git repository (or any of the parent directories):
.git
```
 </div>
 ***

### Fixing a nested `init`

<div class = "help">
What if you forgot to check `git status` and accidentally created a repository nested in another repository?
You can fix this by removing the `.git` file from the inner repository. Take another look at Dracula's code from our last example:

```console
$ cd ~/Desktop   # return to Desktop directory
$ cd planets     # go into planets directory, which is already a Git repository
$ ls -a          # ensure the .git subdirectory is still present in the planets directory
$ mkdir moons    # make a subdirectory planets/moons
$ cd moons       # go into moons subdirectory
$ git init       # make the moons subdirectory a Git repository
$ ls -a          # ensure the .git subdirectory is present indicating we have created a new Git repository
```

Dracula has created two tracking files, `planets/.git` which he should keep, and `planets/moons/.git` which should be removed.
Dracula can do this by running `pwd` which will show the **present working directory** and getting the output:

```console
/Users/vlad/Desktop/planets
```
before running:

```console

> $ rm -rf moons/.git

Reviewer: this is a powerful command and you should explain what rm means and the flags

Author: Good call! Added a warning box

```
</div>

## Tracking changes

>  It might seem like *making* changes and *documenting* them are two different things, but the entire point of version control is to document every change you make.

Reviewer: ```suggestion
 It might seem like **making** changes and **documenting** them are two different things, but the entire point of version control is to document every change you make.
```

 Just like any system that can help back-up or save your work, Git will only be helpful if you use it. In this section we will first teach you

>  - *how* to make a record of your changes,

Reviewer: ```suggestion
 - **how** to make a record of your changes,
```

>  - *where* those records are, and

Reviewer: ```suggestion
 - **where** those records are, and
```

>  - *when* you should be recording that you changed things.

Reviewer: ```suggestion
 - **when** you should be recording that you changed things.
```

	new file:   mars.txt
```

Git now knows that it's supposed to keep track of `mars.txt`,
but it hasn't recorded these changes as a commit yet.
To get it to do that, we need to run one more command. Enter this into the console:

```console
$ git commit -m "Start notes on Mars as a base"
```
The console output will look like this, though your unique identifiers (f22b25e and 100644) may be different.

```console
[main (root-commit) f22b25e] Start notes on Mars as a base
 1 file changed, 1 insertion(+)
 create mode 100644 mars.txt
```

When we run `git commit`,
Git takes everything we have told it to save by using `git add`
and stores a copy permanently inside the special `.git` directory.
This permanent copy is called a **commit**
(or **revision**) and its short identifier is `f22b25e`. Your commit may have another identifier.

We use the `-m` flag (for "message")
to record a short, descriptive, and specific comment that will help us remember later on what we did and why.
If we just run `git commit` without the `-m` option,
Git will launch `nano` (or whatever other editor we configured as `core.editor`)
so that we can write a longer message.

> <div class = "important">

Reviewer: This isn't rendering correctly

**Is there a better way to think of this?** Lots of people have favorite metaphors for how Git works. Before checking out the selection on the next page, do you have any ideas for possible ways to understand the `add` and `commit` process in Git?

### A few metaphors for Git
Metaphors for Git abound, and no metaphor has been great enough to become the "standard" example. If you are looking for a good enough metaphor to help you understand what Git is, here are some to consider:

* A [poetry metaphor](https://www.youtube.com/watch?v=BCQHnlnPusY) video on YouTube ~14 minutes

* A [musical metaphor](https://www.youtube.com/watch?v=S9Do2p4PwtE) video on YouTube ~5 minutes

* A [metaphor about building a house](https://towardsdatascience.com/a-simple-story-to-explain-version-control-to-anyone-5ab4197cebbc) blog post

* A [metaphor about photo albums](https://www.freecodecamp.org/news/git-the-laymans-guide-to-understanding-the-core-concepts/) article

Remember a metaphor doesn't need to be perfect to be helpful. If you find or come up with one that helps you understand the workflow, use it!

> ### Telling Git not to track some files

Reviewer: .DS_Store is the perennial fave for mac users, maybe mention?.  https://gist.github.com/octocat/9257657 might be a good resource to mention here too.

Also.  Anything like a file containing an API key or anything with any level of secrecy should be .gitignored.  Once this information is committed and pushed to GitHub, it remains in history, even if you remove that file in a later commit.  This is a sad thing that happens to a lot of people and warning them about that now will be helpful

Author: @pm0kjp 

GREAT call. Actually this seems important enough to become a learning objective and have a quiz of its own, so I added those.

Author: @pm0kjp 
I made .gitignore into its own section with its own quiz.

> What if we have files that we do not want Git to track for us,

Reviewer: This section has some line breaks in the middle of sentences... I don't see that showing up in the rendered version, but maybe let text wrap?  Unsure of whether this is some sort of weird github artifact that isn't real.

Author: I think it is an artifact from SWCarpentry, fixing it everywhere I see it!

```
$ git add a.dat

The following paths are ignored by one of your .gitignore files:

 a.dat

Use -f if you really want to add them.
```

If we really want to override our ignore settings,
we can use `git add -f` to force Git to add something. For example,
`git add -f a.dat`.
We can also always see the status of ignored files if we want:

```
$ git status --ignored

On branch main
Ignored files:
(use "git add -f <file>..." to include in what will be committed)

a.dat
b.dat
c.dat
results/

nothing to commit, working directory clean
```

For more details on how to use `.gitignore` to include or exclude particular files or folders check out Software Carpentry's [lesson on Ignoring Things](https://swcarpentry.github.io/git-novice/06-ignore/index.html).

### Quiz: Git workflow

What does the `git commit` command do?

 [[ ]] Stages files so that their current version can be recorded.
 [[ ]] Creates a record of the current version of all files in the repository.
 [[X]] Creates a record of the current version of all of the staged files in the repository.
***
<div class = "answer">
The `git commit` command will only create a record of the files that you have *already staged.*

> There are options you can use to automatically stage all files. Entering `git commit -a` or `git commit --all` will automatically stage all files that Git knows about and then commit them. But it will not save any new files that Git doesn't yet know about. In our artwork metaphor, if you started a new drawing on another sheet of paper, `git commit -a` would not put the new drawing into the slot and scan it, but it would scan the current version of every drawing that had previously been scanned.

Reviewer: what artwork metaphor?  I think we removed this, right?

### Initializing the repository

First, let's create a directory folder for our work. Depending on how your computer is organized the `Desktop` on Mac or Linux, or in `C` Drive on Windows machines may be good locations for the new directory folder, but you can put it wherever you like. Dracula's computer has a `Desktop` where he does his work.

```console
$ cd ~/Desktop
$ mkdir planets
$ cd planets
```
The `cd` command means **change directory**, and `mkdir` tells the computer to **make** a **directory** with the name "planets." Once we have a `planets` directory, we use `cd` open that directory.

<div class = "help">
Getting an error message when you type these commands? If you are using a Windows machine, you need to switch over to Git Bash!
</div>

Next we tell Git to make `planets` a [repository](https://swcarpentry.github.io/git-novice/reference.html#repository) (also called a **repo**)
-- a place where Git can store versions of our files:

```console

> $ git init

Author: Good call! I don't think this is just older versions of Git either. Added a details box if people are curious at this point, and also still have the original comment at the end of setting the branch to main.

```

It is important to note that `git init` will create a repository that
includes subdirectories and their files---there is no need to create
separate repositories nested within the `planets` repository, whether
subdirectories are present from the beginning or added later. Also, note
that the creation of the `planets` directory and its initialization as a
repository are completely separate processes.

If we use `ls`, the **list** command, to show the directory's contents,

> it appears that nothing has changed:

Author: now reads "it appears to be empty"

repository in the directory, check the output of `git status`. If it looks
like the following, you are good to go to create a new repository as shown
above:

```console
$ git status

fatal:
Not a git repository (or any of the parent directories):
.git
```
 </div>
 ***

### Fixing a nested `init`

What if you forgot to check `git status` and accidentally created a repository nested in another repository?
You can fix this by removing the `.git` file from the inner repository. Take another look at Dracula's code from our last example:

```console
$ cd ~/Desktop   # return to Desktop directory
$ cd planets     # go into planets directory, which is already a Git repository
$ ls -a          # ensure the .git subdirectory is still present in the planets directory
$ mkdir moons    # make a subdirectory planets/moons
$ cd moons       # go into moons subdirectory
$ git init       # make the moons subdirectory a Git repository
$ ls -a          # ensure the .git subdirectory is present indicating we have created a new Git repository
```

Dracula has created two tracking files, `planets/.git` which he should keep, and `planets/moons/.git` which should be removed.

> Dracula can do this by running `pwd` which will show the **present working directory** and getting the output:

Author: rephrased with neither `ls` nor `pwd`.

```console
/Users/vlad/Desktop/planets
```
before running:

```console
$ rm -rf moons/.git
```
<div class = "warning">
The `rm` command **removes** files, and can be very useful, but the `-rf` flag is very powerful and should used with great caution. The `r` stands for "recursive" and instructs the computer to remove all contents of the file and then remove the file itself.  The `f` is for "force" and means you will not be asked if you are sure you want to delete the files.

Dracula should be absolutely certain he wants to delete `moons/.git` before running this command.
</div>

## Tracking changes

>  It might seem like **making** changes and **documenting** them are two different things, but the entire point of version control is to document every change you make.

Author: removed the entire sentence, this was written shortly after I hung out with a grumpy CS prof friend :)

 Just like any system that can help back-up or save your work, Git will only be helpful if you use it. In this section we will first teach you

 - **how** to make a record of your changes,

 - **where** those records are, and

 - **when** you should be recording that you changed things.

### Tell Git to track a file
Let's follow along with Dracula to add a file to the planets directory with notes about Mars.

First let's make sure we're still in the right directory.

> You should be in the `planets` directory.

Author: introducing pwd here, especially since I'm removing it from earlier in the module

```console
$ cd ~/Desktop/planets
```

Let's create a file called `mars.txt` that contains some notes
about the Red Planet's suitability as a base.
We'll use `nano` to edit the file;
you can use whatever editor you like.
In particular, this does not have to be the `core.editor` you set globally earlier. But remember, the bash command to create or edit a new file will depend on the editor you choose (it might not be `nano`). For a refresher on text editors, check out ["Which Editor?"](https://swcarpentry.github.io/shell-novice/03-create/) in [The Unix Shell lesson](https://swcarpentry.github.io/shell-novice/) by Software Carpentry.

```console
$ nano mars.txt
```

Type the text below into the `mars.txt` file:

```console
Cold and dry, but everything is my favorite color
```

Let's first verify that the file was properly created by running the list command (`ls`):

```console
$ ls

mars.txt
```

> `mars.txt` contains a single line, which we can see by running `cat`:

Author: updated and added a very short `cat` explanation.

```console
$ cat mars.txt

Cold and dry, but everything is my favorite color
```

If we check the status of our project again,
Git tells us that it's noticed the new file:

```console
$ git status

On branch main

No commits yet

Untracked files:
   (use "git add <file>..." to include in what will be committed)

> 	mars.txt

Author: It looks like you can't style things inside a code block because it reads everything verbatim. Which makes sense because that is what we mean by making it a code block. It seems like a significant amount of work to make a new code block environment in css to do this, so I think the answer is no.

nothing added to commit but untracked files present (use "git add" to track)
```

The "untracked files" message means that there's a file in the directory that Git isn't keeping track of.
We can tell Git to track a file using `git add`:

```console
$ git add mars.txt
```

and then use `git status` to check that the right thing happened:

```console
$ git status

On branch main

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

> 	new file:   mars.txt

Author: seems like no :(

```

Git now knows that it's supposed to keep track of `mars.txt`,
but it hasn't recorded these changes as a commit yet.
To get it to do that, we need to run one more command. Enter this into the console:

```console
$ git commit -m "Start notes on Mars as a base"
```
The console output will look like this, though your unique identifiers (f22b25e and 100644) may be different.

```console
[main (root-commit) f22b25e] Start notes on Mars as a base
 1 file changed, 1 insertion(+)
 create mode 100644 mars.txt
```

When we run `git commit`,
Git takes everything we have told it to save by using `git add`
and stores a copy permanently inside the special `.git` directory.
This permanent copy is called a **commit**

> (or **revision**) and its short identifier is `f22b25e`. Your commit may have another identifier.

Author: I don't think I understand what it means for the longer hash to be deterministic well enough to explain it in a concise way here. But I'd love to learn if you have a few minutes to teach me?

<div class = 'help'>
**Where Are My Changes?** If we run `ls` at this point, we will still see just one file called `mars.txt`. That's because Git saves information about files' history in the special `.git` directory mentioned earlier so that our filesystem doesn't become cluttered (and so that we can't accidentally edit or delete an old version).
</div>

Now suppose Dracula adds more information to the file.
(Again, we'll edit with `nano` and then `cat` the file to show its contents;
you may use a different editor, and don't need to `cat`.)

```
$ nano mars.txt
$ cat mars.txt

Cold and dry, but everything is my favorite color
The two moons may be a problem for Wolfman
```

When we run `git status` now,
it tells us that a file it already knows about has been modified:

```
$ git status

On branch main
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   mars.txt

no changes added to commit (use "git add" and/or "git commit -a")
```

The last line is the key phrase:
"no changes added to commit".
We have changed this file,

> but we haven't told Git we will want to save those changes

Author: updated!

(which we do with `git add`)

> nor have we saved them (which we do with `git commit`).

Author: Now talks about telling git to track those changes, and recording the changes.

@pm0kjp please take a look and check for clarity!

  2. The second line tells exactly which versions of the file Git is comparing; `df0654a` and `315bf3a` are unique computer-generated labels for those versions.
  3. The third and fourth lines once again show the name of the file being changed.
  4. The remaining lines are the most interesting, they show us the actual differences and the lines on which they occur. In particular, the `+` marker in the first column shows where we added a line.
</div>

<div class = 'warning'>
If we immediately commit this change using `git commit` what will happen?

Git won't commit because we didn't use `git add` first! Git will give you a message reminding you to use `git add`.

</div>

Let's try adding the file and then committing the changes:

```
$ git add mars.txt
$ git commit -m "Add concerns about effects of Mars' moons on Wolfman"

[main 34961b1] Add concerns about effects of Mars' moons on Wolfman
 1 file changed, 1 insertion(+)
```

Double check that everything worked by running `git status`:

```
$ git status

On branch main

nothing to commit, working directory clean
```

### How often should you `commit`?

> Just like any method of backing up or saving your work, the benefits of using Git for version control are only available if you regularly use them. In our example, we have been committing every time we add or change a line in a single file. That is going to be odious overkill for most projects.

Reviewer: ```suggestion
Just like any method of backing up or saving your work, the benefits of using Git for version control are only available if you regularly use them. In our example, we have been committing every time we add or change a line in a single file. That is going to be odious overkill for most projects.  

Often, when we're working in a project, we change several files in the course of a few minutes or hours of work.  For example, you might change the counts of research subjects in a couple of different places across two or three files related to a research project when you update the weekly numbers.  It's helpful to know that with git, you can change a few files in your repository, then use `git add` to include the changed files (as many as you want) you want to include in a commit, and then do a single commit which explains all of the changes as a whole.
```

Trying to give an example and make it clear that commit can apply across multiple files...

- **Commit before making any changes that could break your code** If your code currently works are you intend to add something new that may break it, make sure you commit so that you will be able to revert to the working version if needed.

- **Commit regularly** Even if you are only making very small changes, don't forget to commit at regular intervals, like before you switch tasks or end your work day.

### Quiz: `add` and `commit`

1. What sequence of commands would track the changes you made to `myFile.txt`?

  a. `$ git add myFile.txt`

  b. `$ git commit myFile.txt`

  c. `$ git add -m "my short, descriptive message"`

  d. `$ git commit -m "my short, descriptive message"`

[( )] First run a, then b
[(X)] First run a, then d
[( )] First run c, then b
[( )] First run c, then d
***
<div class = 'answer'>

> You must run the `git add` to tell Git to stage `myFile.txt`, then you can commit the changes and leave your short, descriptive message. We will discuss how to think of these steps in the [next section](#understanding-the-git-workflow).

Reviewer: This isn't really the next section anymore

Author: fixed!

</div>
***

> 2. With each commit, we want to leave a short, descriptive, message describing the changes we make. Which of these would be the *best* commit message?

Reviewer: ```suggestion
2. With each commit, we want to leave a short, descriptive, message describing the changes we make. Which of these would be the **best** commit message?
```

[( )] "updates"
[( )] "The Martian atmosphere is only 1% oxygen."
[( )] "add a line that the Martian atmosphere is only 1% oxygen"
[(X)] "add comments on Martian atmosphere"
***
<div class = "answer">
The commit message should describe what this commit does. If you want to know the line by line changes you should use the `diff` command. The purpose of the message is so that you know what you changed without having to read all the individual changes, which is especially helpful if you edited many lines or files.
</div>
***

## Telling Git not to track some files

What if we have files that we do not want Git to track for us, like backup files created by our editor or intermediate files created during data analysis?

<div class = "warning">

> It is likely that you will eventually share your repository publicly through [GitHub](github.com). When you do, everything that Git is tracking will become publicly visible. If there is an API key or anything that requires any level of secrecy, make sure you tell Git to ignore it as soon as you create the file.

Reviewer: ```suggestion
It is likely that you will eventually share your repository.  This could be sharing it privately with a team, or publicly through [GitHub](github.com). When you do, everything that Git is tracking will become visible to everyone with access to the repository.  If there is an API key, password, sensitive data, or anything that requires any level of secrecy, make sure you tell Git to ignore it as soon as you create the file.
```

Once information is committed and made public ("pushed" to GitHub), a record of it will persist even if you remove it in a later commit.
</div>

### Creating `.gitignore`

Let's create a few dummy files to practice ignoring:

```
$ mkdir results
$ touch a.dat b.dat c.dat results/a.out results/b.out
```

The `touch` command opens, saves, and closes the files without changing them. Since they didn't previously exist, `touch` is also creating them. Now if we check the status of our repository we see:

```
$ git status

On branch main
Untracked files:
 (use "git add <file>..." to include in what will be committed)

	a.dat
	b.dat
	c.dat
	results/

nothing added to commit but untracked files present (use "git add" to track)
```

> Putting these files under version control would be a waste of disk space. What's worse, having them all listed could distract us from changes that actually matter, so let's tell Git to ignore them.

Reviewer: ```suggestion
In our scenario, these are files we don't want to track.  Putting these files under version control would be a waste of disk space. What's worse, having them all listed could distract us from changes that actually matter, so let's tell Git to ignore them.
```

> We do this by creating a file in the root directory of our project called `.gitignore`. Each line in this file is a rule for a type of file Git should ignore. Use your favorite text editor to make `.gitignore` contain the two lines:

Reviewer: ```suggestion
We do this by creating a file in the root directory of our project called `.gitignore`.   That dot at the beginning is intentional!  The file must be named this exactly for `.gitignore` to work properly.  Each line in this file is a rule for a type of file Git should ignore. Use your favorite text editor to make `.gitignore` contain the two lines:
```

```
*.dat
results/
```

These patterns tell Git to ignore any file whose name ends in `.dat` and everything in the `results` directory. (If any of these files were already being tracked, Git would continue to track them.)

<div class = "learnmore">

> There are many common file types that people frequently ask Git not to track. For example Mac users usually want Git to ignore `.DS_store` an invisible file created by opening a file in Finder. You can even start with [this collection of common configurations](https://gist.github.com/octocat/9257657) for `.gitignore`.

Reviewer: ```suggestion
There are many common file types that people frequently ask Git not to track. For example Mac users usually want Git to ignore `.DS_store`, which is an invisible file that gets created by opening a file in Finder. You can even start with [this collection of common configurations](https://gist.github.com/octocat/9257657) for `.gitignore`.
```

Let's add and commit `.gitignore`:

```
$ git add .gitignore
$ git commit -m "Ignore data files and the results folder."
$ git status

On branch main

nothing to commit, working directory clean
```

As a bonus, using `.gitignore` helps us avoid accidentally adding files to the repository that we don't want to track:

```
$ git add a.dat

The following paths are ignored by one of your .gitignore files:

 a.dat

Use -f if you really want to add them.
```

If we really want to override our ignore settings,
we can use `git add -f` to force Git to add something. For example,
`git add -f a.dat`.
We can also always see the status of ignored files if we want:

```
$ git status --ignored

On branch main
Ignored files:
(use "git add -f <file>..." to include in what will be committed)

a.dat
b.dat
c.dat
results/

nothing to commit, working directory clean
```

<div class = "learnmore">
For more details on how to use `.gitignore` to include or exclude particular files or folders check out Software Carpentry's [lesson on Ignoring Things](https://swcarpentry.github.io/git-novice/06-ignore/index.html).
</div>

### Quiz: `.gitignore`

> Why might you want to tell Git to NOT track a file? Check all that apply.

Reviewer: This rendered for me as a "can only check one" quiz...

[(X)] The file contains secret information like passwords or API keys.
[(X)] The file contains intermediate data that can be recreated by running other code in your repository.
[(X)] The file is automatically generated by your computer and not actually part of your project.
***
<div class = "answer">

> All of these are good reasons to include a file in `.gitignore` but especially the first reason. Keeping secret information like passwords or API keys is essential.

Reviewer: ```suggestion
All of these are good reasons to include a file in `.gitignore`, but especially the first reason. Keeping secret information like passwords or API keys is essential.
```

</div>
***

## Understanding the Git workflow

There were a lot of steps to getting our changes saved in Git! You could memorize that sequence of steps, and that is what [many people do](https://xkcd.com/1597/). However you might remember them better if you understand what each is doing.

> So far we have encountered files that in three states:

Reviewer: ```suggestion
So far we have encountered files in three states:
```

### A few metaphors for Git
Metaphors for Git abound, and no metaphor has been great enough to become the "standard" example. If you are looking for a good enough metaphor to help you understand what Git is, here are some to consider:

* A [poetry metaphor](https://www.youtube.com/watch?v=BCQHnlnPusY) video on YouTube ~14 minutes

* A [musical metaphor](https://www.youtube.com/watch?v=S9Do2p4PwtE) video on YouTube ~5 minutes

* A [metaphor about building a house](https://towardsdatascience.com/a-simple-story-to-explain-version-control-to-anyone-5ab4197cebbc) blog post

* A [metaphor about photo albums](https://www.freecodecamp.org/news/git-the-laymans-guide-to-understanding-the-core-concepts/) article

Remember a metaphor doesn't need to be perfect to be helpful. If you find or come up with one that helps you understand the workflow, use it!

### Quiz: Git workflow

What does the `git commit` command do?

 [[ ]] Stages files so that their current version can be recorded.
 [[ ]] Creates a record of the current version of all files in the repository.
 [[X]] Creates a record of the current version of all of the staged files in the repository.
***
<div class = "answer">

> The `git commit` command will only create a record of the files that you have *already staged.*

Reviewer: ```suggestion
The `git commit` command will only create a record of the files that you have **already staged.**
```

There are options you can use to automatically stage all files. Entering `git commit -a` or `git commit --all` will automatically stage all files that Git knows about and then commit them. But it will not save any new files that Git doesn't yet know about.
</div>
***

## Additional Resources

This module was based on examples from three of [Software Carpentry](https://software-carpentry.org)'s [lessons on using Git](https://swcarpentry.github.io/git-novice/). The story of Dracula and Wolfman's planned trip to Mars is borrowed from them, and more examples are available in the original lessons:

 - Lesson 3: [Creating a Repository](https://swcarpentry.github.io/git-novice/03-create/index.html)
 - Lesson 4: [Tracking Changes](https://swcarpentry.github.io/git-novice/04-changes/index.html)
 - Lesson 6: [Ignoring Things](https://swcarpentry.github.io/git-novice/06-ignore/index.html)

> This module is also the third is a series of modules about Git created by Arcus Education and friends. While we suggest continuing with the next module in the series, [Exploring a Project's History](link/goes/here) the entire collection is available on the [Arcus website](link/goes/here)

Reviewer: Holding this out as a last change.  Also, Arcus is distinct from the R25 work and most people working with this material won't be working with Arcus.  So, maybe leave off the "entire collection" bit for now?

Author: Deleted entire paragraph, though we may later want to add some version of it back in.

To copy the code in the examples, enter everything after the dollar sign "$" into the command line, and then press the enter or return key. If the command has any output, it will not start with a dollar sign, but a new line where you can type after a dollar sign will appear.

```console
$logname
vlad
$
```

The **code block** above tells you that when Dracula types `logname` and presses enter, the output is 'vlad', his username. Try entering `logname` in your own console. There will be other code in front of each `$` describing who and where you are. Since the stuff before the `$` will be different for everyone, it is customary to omit it from example code. We will also omit the `$` after the output.

### Initializing the repository

First, let's create a directory folder for our work. Depending on how your computer is organized the `Desktop` on Mac or Linux, or in `C` Drive on Windows machines may be good locations for the new directory folder, but you can put it wherever you like. Dracula's computer has a `Desktop` where he does his work.

```console
$ cd ~/Desktop
$ mkdir planets
$ cd planets
```

> The `cd` command means **change directory**.  In our first command, we're changing directory to our home directory (that's the `~`, and within that, to the directory called `Desktop`.  If you don't have this directory in your file system, you'll get an error.  That's okay -- just navigate to a different directory you'd like to work in (as long as that directory isn't within a git repository).  The `mkdir` command tells the computer to **make** a **directory** with the name "planets." Once we have a `planets` directory, we use `cd` open that directory.

Reviewer 2: Tiny copy edit comment! Add in a second parenthesis after "...repository)". add a "to" after "...directory, we use `cd`".

Author: added a `)` after the `~` which I think should have had one to begin with!

[( )] It is not necessary to run `git init` inside `moons` but it can't hurt.
[(X)] No, `moons` was already tracked as a subdirectory of `planets` and running `git init` inside `moons` could interfere with version control.
***
<div class = "answer">
No. Dracula does not need to make the `moons` subdirectory a Git repository
because the `planets` repository will track all files, sub-directories, and
subdirectory files under the `planets` directory.  Thus, in order to track
all information about moons, Dracula only needed to add the `moons` subdirectory
to the `planets` directory.

Additionally, Git repositories can interfere with each other if they are "nested":
the outer repository will try to version-control
the inner repository. Therefore, it's best to create each new Git
repository in a separate directory. To be sure that there is no conflicting
repository in the directory, check the output of `git status`. If it looks
like the following, you are good to go to create a new repository as shown
above:

```console
$ git status

fatal:
Not a git repository (or any of the parent directories):
.git
```
 </div>
 ***

### Fixing a nested `init`

> What if you forgot to check `git status` and accidentally created a repository nested in another repository?

Reviewer 2: I don't think you've introduced `git status` yet within this module. It might be worth adding a hint or link to a brief explanation of what `git status` does and why it would have helped prevent the accidental creation of a nested `init`.

This shows us the differences between the current state
of the file and the most recently saved version:

```
$ git diff

diff --git a/mars.txt b/mars.txt
index df0654a..315bf3a 100644
--- a/mars.txt
+++ b/mars.txt
@@ -1 +1,2 @@
 Cold and dry, but everything is my favorite color
+The two moons may be a problem for Wolfman
```

<div class = 'care'>
The output is cryptic because it is meant for a computer, not a human.
It is actually a series of commands for tools like editors and `patch`
telling them how to reconstruct one file given the other. Nevertheless we can gain some useful information
if we break it down into pieces:

  1. The first line tells us that Git is comparing the old and new versions of the file `mars.txt`.
  2. The second line tells exactly which versions of the file Git is comparing; `df0654a` and `315bf3a` are unique computer-generated labels for those versions.
  3. The third and fourth lines once again show the name of the file being changed.
  4. The remaining lines are the most interesting, they show us the actual differences and the lines on which they occur. In particular, the `+` marker in the first column shows where we added a line.
</div>

> <div class = 'warning'>

Reviewer 2: Do you think it would be worth turning this into a quick quiz question? It already fits that format perfectly!

If we immediately commit this change using `git commit` what will happen?

Git won't commit because we didn't use `git add` first! Git will give you a message reminding you to use `git add`.

</div>

Let's try adding the file and then committing the changes:

```
$ git add mars.txt
$ git commit -m "Add concerns about effects of Mars' moons on Wolfman"

[main 34961b1] Add concerns about effects of Mars' moons on Wolfman
 1 file changed, 1 insertion(+)
```

Double check that everything worked by running `git status`:

```
$ git status

On branch main

nothing to commit, working directory clean
```

### How often should you `commit`?

Just like any method of backing up or saving your work, the benefits of using Git for version control are only available if you regularly use them. In our example, we have been committing every time we add or change a line in a single file. That is going to be odious overkill for most projects.  

> Often, when we're working in a project, we change several files in the course of a few minutes or hours of work.  For example, you might change the counts of research subjects in a couple of different places across two or three files related to a research project when you update the weekly numbers.  It's helpful to know that with git, you can change a few files in your repository, then use `git add` to include the changed files (as many as you want) you want to include in a commit, and then do a single commit which explains all of the changes as a whole.

Reviewer 2: This might be a bigger overhaul that we all come back to later, but might it be useful to also include an example of making changes to multiple files in one commit? Even in the later modules, the examples really sitck to one or two files' changes, but that's so often not how people are really using git day to day.

Author: added an example!

Let's create a few dummy files to practice ignoring:

```
$ mkdir results
$ touch a.dat b.dat c.dat results/a.out results/b.out
```

The `touch` command opens, saves, and closes the files without changing them. Since they didn't previously exist, `touch` is also creating them. Now if we check the status of our repository we see:

```
$ git status

On branch main
Untracked files:
 (use "git add <file>..." to include in what will be committed)

	a.dat
	b.dat
	c.dat
	results/

nothing added to commit but untracked files present (use "git add" to track)
```

In our scenario, these are files we don't want to track.  Putting these files under version control would be a waste of disk space. What's worse, having them all listed could distract us from changes that actually matter, so let's tell Git to ignore them.

We do this by creating a file in the root directory of our project called `.gitignore`.   That dot at the beginning is intentional!  The file must be named this exactly for `.gitignore` to work properly.  Each line in this file is a rule for a type of file Git should ignore. Use your favorite text editor to make `.gitignore` contain the two lines:

```
*.dat
results/
```

> These patterns tell Git to ignore any file whose name ends in `.dat` and everything in the `results` directory. (If any of these files were already being tracked, Git would continue to track them.)

Reviewer 2: Maybe a link to commonly used regex's? e.g. https://git-scm.com/docs/gitignore 

Edit: I see the SWC info on ignoring things is linked later. Could be sufficient!

</div>

Once we have created the `.gitignore` file, the output of `git status` is much cleaner:

```
$ git status

On branch main
Untracked files:
 (use "git add <file>..." to include in what will be committed)

	.gitignore

nothing added to commit but untracked files present (use "git add" to track)
```

The only thing Git notices now is the newly-created `.gitignore` file. You might think we wouldn't want to track it, but if we share the repository, everyone we're sharing with will probably want to ignore the same things that we are ignoring.
Let's add and commit `.gitignore`:

```
$ git add .gitignore
$ git commit -m "Ignore data files and the results folder."
$ git status

On branch main

nothing to commit, working directory clean
```

As a bonus, using `.gitignore` helps us avoid accidentally adding files to the repository that we don't want to track:

```
$ git add a.dat

The following paths are ignored by one of your .gitignore files:

 a.dat

Use -f if you really want to add them.
```

If we really want to override our ignore settings,
we can use `git add -f` to force Git to add something. For example,

> `git add -f a.dat`.

Reviewer 2: I actually don't know the answer to this and don't know that it's critical to put here and now but! I was wondering what the .gitignore file would look like if we did do `git add -f a.dat`. Would it just remove the `a.dat`line from .gitignore? Basically, I'm not sure how .gitignore works when you specify a whole type of file (`*.dat`) to ignore but then override for one file of that type (`git add -f a.dat`).

Will follow up if/when I test this myself!

Author: Good question, I have no idea! Let me know what you find out!

Author: Now we know it doesn't change the `.gitignore` file at all.

> We can also always see the status of ignored files if we want:

Reviewer 2: Add an additional blank line space prior to this sentence to separate from the override ignore example.

```
$ git status --ignored

On branch main
Ignored files:
(use "git add -f <file>..." to include in what will be committed)

a.dat
b.dat
c.dat
results/

nothing to commit, working directory clean
```

<div class = "learnmore">
For more details on how to use `.gitignore` to include or exclude particular files or folders check out Software Carpentry's [lesson on Ignoring Things](https://swcarpentry.github.io/git-novice/06-ignore/index.html).
</div>

### Quiz: `.gitignore`

Why might you want to tell Git to NOT track a file? Check all that apply.

[[X]] The file contains secret information like passwords or API keys.
[[X]] The file contains intermediate data that can be recreated by running other code in your repository.
[[X]] The file is automatically generated by your computer and not actually part of your project.
***
<div class = "answer">
All of these are good reasons to include a file in `.gitignore`, but especially the first reason. Keeping secret information like passwords or API keys is essential.
</div>
***

## Understanding the Git workflow

> There were a lot of steps to getting our changes saved in Git! You could memorize that sequence of steps, and that is what [many people do](https://xkcd.com/1597/). However you might remember them better if you understand what each is doing.

Reviewer 2: lol at the xkcd (very professional and helpful comment right here haha)

**Estimated time to completion:**

@estimated_time

**Pre-requisites**

Before you start this module, make sure you

* Know how to access a command line interface (CLI) on your computer.

> * Have Git configured on your computer.

Reviewer: Just a note here to link this to git install and setup modules when those come on board, no action item necessary rn

Author: They are now up, done!

Open a new command line interface on your computer. All exercises in this module will be done in this interface.

<div class = "warning">

> If you are on a Windows machine, make sure you are in the Git Bash application. The Command Prompt (cmd) will not recognize the commands we are using in this module.  

Reviewer: Add a download link to Git download for windows, which includes git bash?  Perhaps https://git-scm.com/download/win.  Indicate that git bash is part of the windows install and maybe have a quick screen shot of what the icon looks like etc.

</div>

If you are not sure if Git is configured on your computer, enter `git config --list` into your interface. The last two lines of output should show your `user.name` and `user.email`. If these are correct you are all ready to start the module!

> If you need to get Git configured on your computer the [Setting Up Git](https://liascript.io/course/?https://raw.githubusercontent.com/arcus/education_modules/rmh-git-setup/git_setup/setting_up_git.md#1) module will help you get set up in about 20 minutes.

Reviewer: ```suggestion
If you need to get Git configured on your computer the [Setting Up Git](https://liascript.github.io/course/?https://raw.githubusercontent.com/arcus/education_modules/main/setting_up_git_mac_and_linux/setting_up_git_mac_and_linux.md#1) module will help you get set up in about 20 minutes if you're using a Mac or Linux machine.
```

Not sure about how to word this while Windows is still left out of the fun, maybe stage this for integration after the Win setup?

Author: Definitely think we should get the Windows installation integrated and then just link to it after this.

> [Pluto](https://commons.wikimedia.org/wiki/File:PIA19873-Pluto-NewHorizons-FlyingPastImage-20150714-transparent.png) /

Reviewer: So the b-maze art link is questionably safe for work.  Since we link to the dracula art, and *that* links to the author, can we just say "Brian Maze" without a link?  I didn't check other artists but maybe same policy?

Author: @pm0kjp 
Eek! Investigating that image further, it isn't so clear to me that it is CC licensed... although having taken it from SWCarpentry it should be?  I really have no idea how to verify that we can use it at all? Any guidance? 

I've updated the others to just be the image and CC license link.

Author: Turns out they don't have a clear record of who/when that image was uploaded, and took it down and created an issue to replace it. So I'm cutting this image out entirely until/unless we make a new one that we can actually use.

Courtesy NASA/JPL-Caltech.
[Mummy](https://commons.wikimedia.org/wiki/File:Mummy_icon_-_Noun_Project_4070.svg)

> $\copyright$; Gilad Fried / [The Noun Project](https://thenounproject.com/) /

Reviewer: ```suggestion
$\copyright$ Gilad Fried / [The Noun Project](https://thenounproject.com/) /
```

[CC BY 3.0](https://creativecommons.org/licenses/by/3.0/deed.en).
[Moon](https://commons.wikimedia.org/wiki/File:Lune_ico.png)

> $\copyright$; Luc Viatour / [https://lucnix.be](https://lucnix.be/) /

Reviewer: ```suggestion
$\copyright$ Luc Viatour / [https://lucnix.be](https://lucnix.be/) /
```

[CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/deed.en).

In the examples, we will see code as if it is written by Vlad Dracula. If you are practicing these steps as you move through this module, your name and email address will appear instead of Dracula's.

To copy the code in the examples, enter everything after the dollar sign "$" into the command line, and then press the enter or return key. If the command has any output, it will not start with a dollar sign, but a new line where you can type after a dollar sign will appear.

> 

Reviewer: Aside: I usually refer to a "code block" with two words, is "codeblock" a specific kind of thing, or just an alternate way of writing it?

Author: nope! just a typo :) Fixed!

$ mkdir planets
$ cd planets
```

> The `cd` command means **change directory**, and `mkdir` tells the computer to **make** a **directory** with the name "planets." Once we have a `planets` directory, we use `cd` open that directory.

Reviewer: even though we mentioned it, maybe have a warning box here with "seeing this error?  If you're on Windows you probably need to switch over to using Git Bash!"

Author: Added as a "help" box
